x-pack/dockerlogbeat/pipelineManager/pipelineManager.go

```diff
@@ -0,0 +1,233 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package pipelinemanager
+
+import (
+	"crypto/sha1"
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/idxmgmt"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/outputs"
+	"github.com/elastic/beats/libbeat/publisher/pipeline"
+	"github.com/elastic/beats/libbeat/publisher/processing"
+	"github.com/pkg/errors"
+	yaml "gopkg.in/yaml.v2"
+)
+
+// containerConfig is the common.Config unpacking type
+type containerConfig struct {
+	Pipeline pipeline.Config        `config:"pipeline"`
+	Output   common.ConfigNamespace `config:"output"`
+}
+
+// Pipeline represents a single pipeline and the count of associated clients
+type Pipeline struct {
+	pipeline *pipeline.Pipeline
+	refCount int
+}
+
+// PipelineManager is a handler into the map of pipelines used by the plugin
+type PipelineManager struct {
+	mu     *sync.Mutex
+	Logger *logp.Logger
+	// pipelines key: config hash
+	pipelines map[string]*Pipeline
+	// clients config: filepath
+	clients map[string]*ClientLogger
+}
+
+// NewPipelineManager creates a new Pipeline map
+func NewPipelineManager(logCfg *common.Config) *PipelineManager {
+	return &PipelineManager{
+		Logger:    logp.NewLogger("dockerlogbeat"),
+		mu:        new(sync.Mutex),
+		pipelines: make(map[string]*Pipeline),
+		clients:   make(map[string]*ClientLogger),
+	}
+}
+
+// CloseClientWithFile closes the client with the associated file
+func (pm *PipelineManager) CloseClientWithFile(file string) error {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+
+	cl, ok := pm.clients[file]
+	if !ok {
+		return fmt.Errorf("No client for file %s", file)
+	}
+
+	// deincrement the ref count
+	hash := cl.pipelineHash
+	pm.pipelines[hash].refCount--
+
+	pm.Logger.Infof("Closing Client first from pipelineManager")
+	err := cl.Close()
+	if err != nil {
+		return errors.Wrap(err, "error closing client")
+	}
+
+	if pm.pipelines[hash].refCount < 1 {
+		pm.Logger.Infof("Pipeline  closing")
+		pm.pipelines[hash].pipeline.Close()
+		delete(pm.pipelines, hash)
+	}
+
+	return nil
+}
+
+// AddNewPipeline adds a new pipeline with the config to the map
+func (pm *PipelineManager) AddNewPipeline(logOptsConfig map[string]string, name, file string) error {
```

Comment:
I don't see this method being used. Better not export symbols you don't need.

Reply:
Nice catch!